### PR TITLE
Add some support to the dashboad for list datasets

### PR DESCRIPTION
**Pull request by mbirtwell**
# ARTIQ Pull Request

## Description of Changes

Allows viewing small list datasets and editing them. It's not particularly clever it just shows the pyon and allows the user to enter new pyon in the edit dialog. It doesn't have the same checking as when creating datasets, but it's good enough for our use case. 

## Type of Changes

|   | Type |
| ------------- | ------------- |
| ✓  | :sparkles: New feature |

### All Pull Requests

- [x] Use correct spelling and grammar.
- [x] Check the copyright situation of your changes and sign off your patches (`git commit --signoff`, see [copyright](../CONTRIBUTING.rst#copyright-and-sign-off)).

### Code Changes

- [x] Run `flake8` to check code style (follow PEP-8 style). `flake8` has issues with parsing Migen/gateware code, ignore as necessary.
- [x] Test your changes or have someone test them. Mention what was tested and how.

### Git Logistics

- [x] Split your contribution into logically separate changes (`git rebase --interactive`). Merge/squash/fixup commits that just fix or amend previous commits. Remove unintended changes & cleanup. See [tutorial](https://www.atlassian.com/git/tutorials/rewriting-history/git-rebase).
- [x] Write short & meaningful commit messages. Review each commit for messages (`git show`). Format:
  ```
  topic: description. < 50 characters total.
  
  Longer description. < 70 characters per line
  ```

### Licensing

See [copyright & licensing for more info](https://github.com/m-labs/artiq/blob/master/CONTRIBUTING.rst#copyright-and-sign-off).
ARTIQ files that do not contain a license header are copyrighted by M-Labs Limited and are licensed under LGPLv3+.

### CODE REVIEW

**Comment by mbirtwell** [comment]
> > It doesn't have the same checking as when creating datasets
> 
> Why not?

There were two ways of doing checking: the `Creator` which does a pyon parse on every key press and the `Editor` which relied on creating a control which only accepted valid inputs. Adding the `PyonEditor` didn't neatly fit in to either of those ways of working so I skipped doing it nicely, but did just enough that it worked well in practice.

File: artiq/dashboard/datasets.py [hunk 1]
```diff
@@ -49,10 +49,16 @@ def __init__(self, parent, dataset_ctl, key, value):
         buttons.rejected.connect(self.reject)
 
     def accept(self):
-        newkey = self.name_widget.text()
-        value = self.initial_type(self.get_edit_widget_value())
-        asyncio.ensure_future(rename(self.key, newkey, value, self.dataset_ctl))
-        QtWidgets.QDialog.accept(self)
+        try:
+            newkey = self.name_widget.text()
```
[the remark this answers is not in the document]
Comment (mbirtwell) [reply artiq/dashboard/datasets.py:53]: I think it's quite nice it being in the try. What's the issue here? That if it fails then we get an error message that's not quite tailored to this function failing. I'd argue that that's better than getting no error message, which is what happens now.

**Comment by mbirtwell** [comment]
> Unfortunately this attitude results in code that is worse than much of the existing one. We can add this as a `experimental-features` patch if not doable nicely.

I don't think this has adversely affected the code quality significantly, what I was drawing your attention to was an inconsistency in the UI. Upon reflection though that inconsistency already existed, and this builds upon it. The inconsistency arises because the dialogs for creating and editing datasets work in two different ways. This gives rise to the situation where you can create list datasets but couldn't edit them, and you can't change the type of an existing dataset, for example attempting to change a dataset from int to float currently silently drops the fractional part, which I find quite unpleasant.

I think a better way of resolving this would be to combine the create and edit dialogs, add a type drop down, when opening it as create default to pyon, so if the user chooses not to interact with the drop down it behaves as it does today. Changing the drop down can change the type of the value control. For edit the drop down can default to the current type similarly to how we select a different edit widget subclass at the moment.

**Comment by mbirtwell** [comment]
I'm a bit in the middle of a project at the moment, maybe in a couple of weeks I'll be able to carve out some time to do some artiq things. I'm aware that I also have a patch for default gateway configuration on kasli that I never tested because our test hardware was in a strange state when wrote it. So maybe I can knock this and that off then.

**Comment by thomasfire** [comment]
Moved to #2001

**Pull request closed without merging**